mitodl/ocw-hugo-themes
# offline course site theme

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/308

#### What's this PR do?
This PR adds a new theme called `course-offline`. The purpose of this theme is to provide a means of rendering course sites that can be browsed locally from any filesystem. This is accomplished by overriding some partials that are responsible for rendering URLs:

 - `course_home_page_url.html` - This returns the URL to the course home page.
 - `get_destination.html` - This is used in `layouts/_default/_markup/render-link.html` which overrides link rendering from markdown links.
 - `nav_url.html` - This renders URLs in the left nav which comes from the menus configuration file in a content repo, 
 - `resource_url.html` - This is called by various other partials to render URLs to resource pages
 - `webpack_url.html` - This is used to render URLs to resources generated by the Webpack build

Another partial was added in the offline theme called `path_to_root.html`. This partial's job is to take a Hugo Page context and generate a string that is a series of `../` stepping backwards to the root of the site. In some of the above partial overrides, this is called to construct a URL relative to the site root. Another consideration with browsing a site from your local filesystem is that if you hit a "pretty" URL that ends in a folder name, for example `https://ocw.mit.edu/courses/res-8-007-cosmic-origin-of-the-chemical-elements-fall-2019/pages/instructor-insights/`, you will be greeted with a file browser showing you the contents of the folder. For this reason, all URLs need to end in `/index.html`. The overridden partials in `course-offline` make sure that this happens.

#### How should this be manually tested?
This should be tested in conjunction with https://github.com/mitodl/ocw-studio/pull/1453 and https://github.com/mitodl/ocw-hugo-projects/pull/213, however there is some additional testing that should be done outside of the context of the mass build in Concourse. While `course-offline` only has to override a few partials to accomplish having a site be browsable offline, a couple of those partials like `resource_url.html` and `webpack_url.html` were modified to take an additional context argument for finding the path back to root in `course-offline`.  The calls to these partials everywhere they were used had to be updated to pass this argument. The thing to keep in mind when testing here in `ocw-hugo-themes` directly is that the goal is to make sure that the normal theme hasn't been affected. Testing the offline `mass-build-sites` pipeline in the above `ocw-studio` PR will test the actual `course-offline` theme.

 - If you've never run a Hugo site from this repo locally before, read the readme to learn about how to do so. Make sure you have `ocw-www` and at least one course cloned from `mitocwcontent` as well as `ocw-hugo-projects` on the `cg/offline-course-site-2` branch and set these env variables:
```
WWW_CONTENT_PATH=/path/to/mitocwcontent/ocw-www/
COURSE_CONTENT_PATH=/path/to/mitocwcontent/
OCW_TEST_COURSE=<any test course from mitocwcontent>
WWW_HUGO_CONFIG_PATH=/path/to/ocw-hugo-projects/ocw-www/config.yaml
COURSE_HUGO_CONFIG_PATH=/path/to/ocw-hugo-projects/ocw-course/config.yaml
RESOURCE_BASE_URL=https://ocw.mit.edu/
STATIC_API_BASE_URL=https://ocw.mit.edu/
```
 - Spin up `ocw-ww` with `npm run start:www`
 - Visit http://localhost:3000 and make sure you see the OCW home page. You should see no errors in the console and all the images, CSS, etc. should load properly. The links to courses in the carousel won't work, but they should all look as they would on the production site i.e. `http://localhost:3000/courses/res-8-007-cosmic-origin-of-the-chemical-elements-fall-2019`
 - Spin up a course site with `npm run start:course`
 - Visit http://localhost:3000 and make sure you see the course home page. Test the following:
   - Check the console and make sure there are no errors
   - Click nav links to each section and make sure you are brought to the proper page
   - Visit http://localhost:300/resources, find a PDF and click the link
   - Make sure you are brought to the PDF page and the PDF displays properly


#### Where should the reviewer start?
https://github.com/mitodl/ocw-studio/pull/1453 should likely be looked at first, followed by this PR and then https://github.com/mitodl/ocw-hugo-projects/pull/213.

#### Any background context you want to provide?
The `ocw-studio` PR's job is to create a method of pushing up a pipeline to Concourse to run a `mass-build-sites` with this `offline-course` theme and generate an offline build of every course in `ocw-studio`. The `ocw-hugo-projects` PR's purpose is to add an offline config for the course theme.
